src/index/utils.rs

```diff
@@ -23,3 +28,53 @@ pub unsafe fn from_datum(datum: pgrx::pg_sys::Datum) -> DynamicVector {
     }
     vector
 }
+
+pub unsafe fn options_from_list(l: &List<*mut c_void>) -> HashMap<String, String> {
+    l.iter()
+        .map(|m| {
+            let e = *m as *mut pgrx::pg_sys::DefElem;
+            let defname = CStr::from_ptr((*e).defname).to_str().unwrap().to_owned();
+            let defvalue = CStr::from_ptr(pgrx::pg_sys::defGetString(e))
+                .to_str()
+                .unwrap()
+                .to_owned();
+            (defname, defvalue)
+        })
+        .collect()
+}
+
+pub unsafe fn downcast_list<T: Enlist>(l: *mut pgrx::pg_sys::List) -> List<T> {
```

Comment:
It takes ownership. If there are any panics in the hook, it'll be unsound.

Reply:
Finally, I have added a warning for pg12.
It is much complex to exclude pg12 for this feature as our CICD need to refactor to fit it.

Reply:
**It's unsound in all PG versions.**

Reply:
Fixed.